File: src/cost_modelling/calculator.py
"""Core cost calculation engine for ML training estimation.

This module implements the FLOPs-based cost estimation model for training
machine learning models on GPU instances.
"""

from src.cost_modelling.gpu_specs import get_storage_cost


def calculate_training_flops(
    parameter_count: int,
    training_tokens: int,
    architecture: str = "transformer",
    epochs: float = 1.0,
    gradient_checkpointing: bool = False,
) -> float:
    """Calculate total FLOPs required for training.
    
    Uses the standard formula: C = 6 × N × D  (Kaplan et al. 2020,
    "Scaling Laws for Neural Language Models", https://arxiv.org/abs/2001.08361)

    where:
    - N = number of parameters
    - D = total tokens seen = training_tokens × epochs
    - 6 = forward pass (2N) + backward pass (4N) per token

    With gradient checkpointing enabled, the forward pass is recomputed during
    the backward pass, increasing the multiplier from 6 to ~8:
    - C = 8 × N × D  (see Bahdanau, "The FLOPs Calculus of Language Model Training")

    Note on embedding parameters: strictly, N should count only non-embedding
    parameters (Kaplan et al.), as embedding lookups are memory reads with
    negligible FLOPs. For typical models the embedding table is < 5% of total
    parameters so the effect on the estimate is minor.
    
    Args:
        parameter_count: Total number of model parameters
        training_tokens: Number of tokens in the dataset (per epoch)
        architecture: Model architecture type
        epochs: Number of passes through the dataset (default 1.0)
        gradient_checkpointing: Whether activation checkpointing is enabled
        
    Returns:
        Total FLOPs required
    """
    # Architecture-specific base multipliers (forward + backward, no checkpointing)
    # Transformer (6×) is sourced from Kaplan et al. 2020 / Chinchilla 2022.
    # CNN and RNN multipliers are rough approximations; results may vary by architecture.
    base_multipliers = {
        "transformer": 6.0,
        "cnn": 4.0,
        "rnn": 8.0,
        "vit": 6.0,       # Vision Transformer — same derivation as decoder transformer
        "diffusion": 6.5,  # Slightly higher due to noise-prediction overhead
    }

    multiplier = base_multipliers.get(architecture.lower(), 6.0)

    # Gradient checkpointing recomputes activations in the backward pass,
    # adding an extra ~forward pass worth of FLOPs: 2N → multiplier increases by 2
    if gradient_checkpointing:
        multiplier += 2.0

    total_tokens = training_tokens * epochs
    total_flops = multiplier * parameter_count * total_tokens
    
    return total_flops


def estimate_compute_cost(
    total_flops: float,
    peak_flops_per_gpu: float,
    mfu: float,
    total_gpus: int,
    num_instances: int,
    hourly_cost: float,
) -> tuple[float, float, float, float]:
    """Estimate wall-clock time, GPU-hours, and compute cost for a training run.

    Args:
        total_flops: Total FLOPs required for the run
        peak_flops_per_gpu: Effective peak FLOPs/s per GPU for the chosen precision
        mfu: Model FLOPs Utilization (0.0–1.0)
        total_gpus: Total number of GPUs across all instances
        num_instances: Number of instances (for cost calculation)
        hourly_cost: Per-instance hourly cost in USD

    Returns:
        Tuple of (wall_clock_hours, gpu_hours, compute_cost, wall_clock_days)
    """
    effective_cluster_flops = peak_flops_per_gpu * mfu * total_gpus
    wall_clock_seconds = total_flops / effective_cluster_flops
    wall_clock_hours = wall_clock_seconds / 3600
    gpu_hours = wall_clock_hours * total_gpus
    compute_cost = wall_clock_hours * hourly_cost * num_instances
    wall_clock_days = wall_clock_hours / 24
    return wall_clock_hours, gpu_hours, compute_cost, wall_clock_days


# Memory bytes per parameter by training regime:
# QLoRA: 4-bit quantized base (0.5 bytes) + fp16/fp32 adapter optimizer states
# LoRA:  fp16 frozen base (2 bytes) + fp16/fp32 adapter optimizer states
# Full / Pre-training: fp16 weights + fp32 Adam states (m+v) + fp32 master copy ≈ 16 bytes
BYTES_PER_PARAM_QLORA_BASE = 0.5
BYTES_PER_PARAM_LORA_BASE  = 2
BYTES_PER_PARAM_TRAINABLE  = 16   # fp16 weights + fp32 Adam states + gradient buffer
BYTES_PER_PARAM_FULL_FT    = 16   # same as TRAINABLE for full fine-tuning / pre-training

# Activation memory per token per layer (bf16 transformer):
# 4 tensors (QKV projections, attention scores, MLP intermediate, residual) × 2 bytes (bf16)
ACTIVATION_BYTES_PER_TOKEN_PER_LAYER = 4 * 2


def estimate_gpu_memory_gb(
    effective_params: int,
    trainable_params: int,
    ft_method: str | None,
    total_gpus: int,
    rl_multiplier: int,
    d_model: int,
    num_layers: int,
    seq_len: int,
    batch_size: int,
    gradient_checkpointing: bool,
) -> tuple[float, float, float]:
    """Estimate GPU memory requirements for a training run.

    Args:
        effective_params: Total (or active, for MoE) model parameter count
        trainable_params: Parameters that receive gradient updates
        ft_method: "QLoRA", "LoRA", "Full Fine-Tuning", or None (pre-training)
        total_gpus: Total GPUs (memory is distributed across them)
        rl_multiplier: Extra model copies for RL (2 for DPO, 4 for PPO)
        d_model: Hidden dimension (0 to skip activation estimate)
        num_layers: Number of transformer layers (0 to skip activation estimate)
        seq_len: Sequence length in tokens (0 to skip activation estimate)
        batch_size: Global batch size
        gradient_checkpointing: Whether activation recomputation is enabled

    Returns:
        Tuple of (weights_gb, activation_gb, memory_per_gpu_gb)
    """
    if ft_method == "QLoRA":
        model_memory_bytes = (
            effective_params * BYTES_PER_PARAM_QLORA_BASE
            + trainable_params * BYTES_PER_PARAM_TRAINABLE
        )
    elif ft_method == "LoRA":
        model_memory_bytes = (
            effective_params * BYTES_PER_PARAM_LORA_BASE
            + trainable_params * BYTES_PER_PARAM_TRAINABLE
        )
    else:
        model_memory_bytes = effective_params * BYTES_PER_PARAM_FULL_FT

    model_memory_bytes *= rl_multiplier
    weights_gb = model_memory_bytes / total_gpus / 1e9

    if d_model and num_layers and seq_len:
        if gradient_checkpointing:
            # With checkpointing, only one layer of activations is live at a time
            activation_bytes = (
                batch_size * seq_len * d_model * ACTIVATION_BYTES_PER_TOKEN_PER_LAYER
            )
        else:
            activation_bytes = (
                batch_size * seq_len * d_model * num_layers
                * ACTIVATION_BYTES_PER_TOKEN_PER_LAYER
            )
    else:
        activation_bytes = 0

    activation_gb = activation_bytes / total_gpus / 1e9
    memory_per_gpu_gb = weights_gb + activation_gb
    return weights_gb, activation_gb, memory_per_gpu_gb


def calculate_lora_trainable_params(
    ft_method: str,
    base_params: int,
    target_modules: list[str],
    d_model: int,
    num_layers: int,
    lora_rank: int,
    architecture: dict,
) -> int | None:
    """Calculate the number of trainable parameters for a fine-tuning run.

    For Full Fine-Tuning returns base_params.

    For LoRA / QLoRA with a known architecture dict (non-empty), uses per-module
    dimensions to account for GQA and non-square projection matrices.

    For custom models (architecture={}), falls back to the uniform approximation:
        2 × rank × d_model × num_modules × num_layers

    Returns None when required inputs (target_modules, d_model, num_layers,
    lora_rank) are missing or zero.
    """
    if ft_method == "Full Fine-Tuning":
        return base_params

    if not (target_modules and d_model and num_layers and lora_rank):
        return None

    if architecture:
        _d        = architecture.get("d_model", d_model)
        _nh       = architecture.get("num_heads", 0)
        _nkv      = architecture.get("num_kv_heads", _nh)
        _head_dim = _d // _nh if _nh else _d
        _kv_dim   = _nkv * _head_dim
        _ffn      = architecture.get("ffn_intermediate", _d * 4)
        module_dims = {
            "q_proj":    (_d, _d),
            "k_proj":    (_d, _kv_dim),
            "v_proj":    (_d, _kv_dim),
            "o_proj":    (_d, _d),
            "up_proj":   (_d, _ffn),
            "down_proj": (_ffn, _d),
        }
        return sum(
            lora_rank * (in_d + out_d)
            for mod in target_modules
            for in_d, out_d in [module_dims.get(mod, (_d, _d))]
        ) * num_layers
    else:
        # Uniform d_model approximation for custom models
        return len(target_modules) * 2 * lora_rank * d_model * num_layers


# Bytes per parameter in a mixed-precision checkpoint:
# fp16 weights = 2 bytes
# fp32 Adam first moment (m) = 4 bytes
# fp32 Adam second moment (v) = 4 bytes
# fp32 master weight copy = 4 bytes
# Total = 14 bytes/param
BYTES_PER_PARAM_CHECKPOINT = 14


def calculate_checkpoint_storage_tb(
    checkpoint_params: int,
    num_checkpoints: int,
    num_training_runs: int,
    num_hp_trials: int,
    num_ablations: int,
) -> float:
    """Calculate total S3 storage required for model checkpoints in TB.

    Full training runs save num_checkpoints each; HP trials and ablations
    save one final checkpoint each.

    Args:
        checkpoint_params: Parameters saved per checkpoint (full model or LoRA adapters)
        num_checkpoints: Checkpoints saved per full training run
        num_training_runs: Number of full training runs
        num_hp_trials: Number of hyperparameter tuning trials
        num_ablations: Number of ablation studies

    Returns:
        Total checkpoint storage in terabytes
    """
    total_bytes = (
        checkpoint_params * BYTES_PER_PARAM_CHECKPOINT * num_checkpoints * num_training_runs
        + checkpoint_params * BYTES_PER_PARAM_CHECKPOINT * num_hp_trials
        + checkpoint_params * BYTES_PER_PARAM_CHECKPOINT * num_ablations
    )
    return total_bytes / 1e12


def calculate_project_compute_cost(
    single_run_cost: float,
    num_training_runs: int,
    num_hp_trials: int,
    hp_fraction: float,
    num_ablations: int,
    ablation_fraction: float,
) -> float:
    """Calculate total GPU compute cost across all training runs, HP trials, and ablations.

    Args:
        single_run_cost: Compute cost of one full training run in USD
        num_training_runs: Number of full training runs
        num_hp_trials: Number of hyperparameter tuning trials
        hp_fraction: Fraction of a full run's cost per HP trial
        num_ablations: Number of ablation studies
        ablation_fraction: Fraction of a full run's cost per ablation

    Returns:
        Total compute cost in USD
    """
    return (
        single_run_cost * num_training_runs
        + single_run_cost * hp_fraction * num_hp_trials
        + single_run_cost * ablation_fraction * num_ablations
    )


def calculate_storage_cost(
    dataset_size_tb: float,
    storage_duration_months: float = 1.0,
    storage_class: str = "standard"
) -> float:
    """Calculate S3 storage cost for training data.
    
    Args:
        dataset_size_tb: Dataset size in terabytes
        storage_duration_months: How long to store the data
        storage_class: S3 storage class
        
    Returns:
        Total storage cost in USD
    """
    cost_per_tb_month = get_storage_cost(storage_class)
    return dataset_size_tb * cost_per_tb_month * storage_duration_months

File: lib/engine/calculator.ts
import { getStorageCost } from "./gpuSpecs";
import type { ComputeCost, FineTuningMethod, GpuMemory, ModelArchitecture } from "./types";

/**
 * Memory bytes per parameter by training regime:
 *   QLoRA: 4-bit quantized base (0.5 bytes) + fp16/fp32 adapter optimizer states
 *   LoRA:  fp16 frozen base (2 bytes) + fp16/fp32 adapter optimizer states
 *   Full / Pre-training: fp16 weights + fp32 Adam states (m+v) + fp32 master copy
 */
export const BYTES_PER_PARAM_QLORA_BASE = 0.5;
export const BYTES_PER_PARAM_LORA_BASE = 2;
export const BYTES_PER_PARAM_TRAINABLE = 16;
export const BYTES_PER_PARAM_FULL_FT = 16;

/** 4 tensors (QKV projections, attention scores, MLP intermediate, residual) x 2 bytes (bf16). */
export const ACTIVATION_BYTES_PER_TOKEN_PER_LAYER = 4 * 2;

/**
 * Bytes per parameter in a mixed-precision checkpoint:
 * fp16 weights (2) + fp32 Adam m (4) + fp32 Adam v (4) + fp32 master copy (4) = 14.
 */
export const BYTES_PER_PARAM_CHECKPOINT = 14;

const BASE_FLOPS_MULTIPLIERS: Record<string, number> = {
  transformer: 6.0,
  cnn: 4.0,
  rnn: 8.0,
  vit: 6.0,
  diffusion: 6.5
};

export function getFlopsMultiplier(architecture: string, gradientCheckpointing: boolean): number {
  let multiplier = BASE_FLOPS_MULTIPLIERS[architecture.toLowerCase()] ?? 6.0;
  if (gradientCheckpointing) multiplier += 2.0;
  return multiplier;
}

export function calculateTrainingFlops(
  parameterCount: number,
  trainingTokens: number,
  architecture = "transformer",
  epochs = 1.0,
  gradientCheckpointing = false
): number {
  const multiplier = getFlopsMultiplier(architecture, gradientCheckpointing);
  return multiplier * parameterCount * trainingTokens * epochs;
}

export function estimateComputeCost(
  totalFlops: number,
  peakFlopsPerGpu: number,
  mfu: number,
  totalGpus: number,
  numInstances: number,
  hourlyCost: number
): ComputeCost {
  const effectiveClusterFlops = peakFlopsPerGpu * mfu * totalGpus;
  // Python raises ZeroDivisionError here; returning zeroes keeps the UI from rendering
  // "$Infinity" if component state reaches a transient the widget minimums don't allow.
  if (effectiveClusterFlops === 0) {
    return { wall_clock_hours: 0, gpu_hours: 0, compute_cost: 0, wall_clock_days: 0 };
  }
  const wallClockSeconds = totalFlops / effectiveClusterFlops;
  const wallClockHours = wallClockSeconds / 3600;
  return {
    wall_clock_hours: wallClockHours,
    gpu_hours: wallClockHours * totalGpus,
    compute_cost: wallClockHours * hourlyCost * numInstances,
    wall_clock_days: wallClockHours / 24
  };
}

const safeCount = (value: number): number =>
  Number.isFinite(value) ? Math.max(0, Math.trunc(value)) : 0;

export function solveForParameterCount(
  computeBudgetUsd: number,
  trainingTokens: number,
  peakFlopsPerGpu: number,
  mfu: number,
  totalGpus: number,
  numInstances: number,
  hourlyCost: number,
  architecture = "transformer",
  epochs = 1.0,
  gradientCheckpointing = false
): number {
  const multiplier = getFlopsMultiplier(architecture, gradientCheckpointing);
  const totalTokens = trainingTokens * epochs;
  const effectiveClusterFlops = peakFlopsPerGpu * mfu * totalGpus;

  if (totalTokens <= 0 || hourlyCost <= 0 || numInstances <= 0) return 0;

  const parameterCount =
    (computeBudgetUsd * effectiveClusterFlops * 3600) /
    (multiplier * totalTokens * hourlyCost * numInstances);
  return safeCount(parameterCount);
}

export function solveForTrainingTokens(
  computeBudgetUsd: number,
  parameterCount: number,
  peakFlopsPerGpu: number,
  mfu: number,
  totalGpus: number,
  numInstances: number,
  hourlyCost: number,
  architecture = "transformer",
  epochs = 1.0,
  gradientCheckpointing = false
): number {
  const multiplier = getFlopsMultiplier(architecture, gradientCheckpointing);
  const effectiveClusterFlops = peakFlopsPerGpu * mfu * totalGpus;

  if (parameterCount <= 0 || hourlyCost <= 0 || numInstances <= 0) return 0;

  const totalTokens =
    (computeBudgetUsd * effectiveClusterFlops * 3600) /
    (multiplier * parameterCount * hourlyCost * numInstances);
  return safeCount(totalTokens / Math.max(epochs, 1.0));
}

export function estimateGpuMemoryGb(
  effectiveParams: number,
  trainableParams: number,
  ftMethod: FineTuningMethod,
  totalGpus: number,
  rlMultiplier: number,
  dModel: number,
  numLayers: number,
  seqLen: number,
  batchSize: number,
  gradientCheckpointing: boolean
): GpuMemory {
  let modelMemoryBytes: number;
  if (ftMethod === "QLoRA") {
    modelMemoryBytes =
      effectiveParams * BYTES_PER_PARAM_QLORA_BASE + trainableParams * BYTES_PER_PARAM_TRAINABLE;
  } else if (ftMethod === "LoRA") {
    modelMemoryBytes =
      effectiveParams * BYTES_PER_PARAM_LORA_BASE + trainableParams * BYTES_PER_PARAM_TRAINABLE;
  } else {
    modelMemoryBytes = effectiveParams * BYTES_PER_PARAM_FULL_FT;
  }

  modelMemoryBytes *= rlMultiplier;
  if (totalGpus === 0) {
    return { weights_gb: 0, activation_gb: 0, memory_per_gpu_gb: 0 };
  }
  const weightsGb = modelMemoryBytes / totalGpus / 1e9;

  let activationBytes = 0;
  if (dModel && numLayers && seqLen) {
    activationBytes = gradientCheckpointing
      ? // With checkpointing only one layer of activations is live at a time
        batchSize * seqLen * dModel * ACTIVATION_BYTES_PER_TOKEN_PER_LAYER
      : batchSize * seqLen * dModel * numLayers * ACTIVATION_BYTES_PER_TOKEN_PER_LAYER;
  }

  const activationGb = activationBytes / totalGpus / 1e9;
  return {
    weights_gb: weightsGb,
    activation_gb: activationGb,
    memory_per_gpu_gb: weightsGb + activationGb
  };
}

export function calculateLoraTrainableParams(
  ftMethod: string,
  baseParams: number,
  targetModules: string[],
  dModel: number,
  numLayers: number,
  loraRank: number,
  architecture: ModelArchitecture
): number | null {
  if (ftMethod === "Full Fine-Tuning") return baseParams;

  if (!(targetModules.length && dModel && numLayers && loraRank)) return null;

  if (architecture && Object.keys(architecture).length > 0) {
    const d = architecture.d_model ?? dModel;
    const numHeads = architecture.num_heads ?? 0;
    const numKvHeads = architecture.num_kv_heads ?? numHeads;
    // Floor division, matching Python's `//`. Using `/` silently changes k/v_proj sizing.
    const headDim = numHeads ? Math.floor(d / numHeads) : d;
    const kvDim = numKvHeads * headDim;
    const ffn = architecture.ffn_intermediate ?? d * 4;
    const moduleDims: Record<string, [number, number]> = {
      q_proj: [d, d],
      k_proj: [d, kvDim],
      v_proj: [d, kvDim],
      o_proj: [d, d],
      up_proj: [d, ffn],
      down_proj: [ffn, d]
    };
    const total = targetModules.reduce((sum, mod) => {
      const [inD, outD] = moduleDims[mod] ?? [d, d];
      return sum + loraRank * (inD + outD);
    }, 0);
    return total * numLayers;
  }

  // Uniform d_model approximation for custom models
  return targetModules.length * 2 * loraRank * dModel * numLayers;
}

export function calculateCheckpointStorageTb(
  checkpointParams: number,
  numCheckpoints: number,
  numTrainingRuns: number,
  numHpTrials: number,
  numAblations: number
): number {
  const totalBytes =
    checkpointParams * BYTES_PER_PARAM_CHECKPOINT * numCheckpoints * numTrainingRuns +
    checkpointParams * BYTES_PER_PARAM_CHECKPOINT * numHpTrials +
    checkpointParams * BYTES_PER_PARAM_CHECKPOINT * numAblations;
  return totalBytes / 1e12;
}

export function calculateProjectComputeCost(
  singleRunCost: number,
  numTrainingRuns: number,
  numHpTrials: number,
  hpFraction: number,
  numAblations: number,
  ablationFraction: number
): number {
  return (
    singleRunCost * numTrainingRuns +
    singleRunCost * hpFraction * numHpTrials +
    singleRunCost * ablationFraction * numAblations
  );
}

/** S3 storage cost for training data. Throws on an unknown storage class. */
export function calculateStorageCost(
  datasetSizeTb: number,
  storageDurationMonths = 1.0,
  storageClass = "standard"
): number {
  const costPerTbMonth = getStorageCost(storageClass);
  return datasetSizeTb * costPerTbMonth * storageDurationMonths;
}
